libraries/AdaptiveExpressions/ExpressionFunctions.cs

```diff
@@ -3118,21 +3118,36 @@ private static IEnumerable<object> Flatten(IEnumerable<object> list, int dept)
                 // String
                 new ExpressionEvaluator(
                     ExpressionType.Concat,
-                    Apply(
+                    ApplySequence(
                         args =>
                         {
-                            var builder = new StringBuilder();
-                            foreach (var arg in args)
+                            var firstItem = args[0];
+                            var secondItem = args[1];
+                            var isFirstList = TryParseList(firstItem, out var firstList);
+                            var isSecondList = TryParseList(secondItem, out var secondList);
+
+                            if (firstItem == null && secondItem == null)
                             {
-                                if (arg != null)
-                                {
-                                    builder.Append(arg.ToString());
-                                }
+                                return null;
+                            }
+                            else if (firstItem == null && isSecondList)
+                            {
+                                return secondList;
+                            }
+                            else if (secondItem == null && isFirstList)
+                            {
+                                return firstList;
+                            }
+                            else if (isFirstList && isSecondList)
+                            {
+                                return firstList.OfType<object>().Concat(secondList.OfType<object>()).ToList();
+                            }
+                            else
+                            {
+                                return $"{firstItem?.ToString()}{secondItem?.ToString()}";
```

Comment:
>return $"{firstItem?.ToString()}{secondItem?.ToString()}"; [](start = 32, length = 58)

Should not use ToString().  If you can't convert to string type, then it is an error.  Even things like number are ambiguous on how they get converted to strings depending on locale.  For other kinds of object, toString might be diagnostic.  This will also be much less efficient than the previous implementation since it does not use a string builder.  Why not extend ApplySequence to use an accumulator which can use a string builder?